File: packages/aegis-ledger-sdk/src/canonical.ts
import { createHash } from "node:crypto";

/** Stable JSON serialization (sorted object keys, no chain.event_hash). */
export function canonicalize(value: unknown): string {
  return JSON.stringify(sortValue(value));
}

function sortValue(value: unknown): unknown {
  if (Array.isArray(value)) {
    return value.map(sortValue);
  }
  if (value !== null && typeof value === "object") {
    const obj = value as Record<string, unknown>;
    const sorted: Record<string, unknown> = {};
    for (const key of Object.keys(obj).sort()) {
      sorted[key] = sortValue(obj[key]);
    }
    return sorted;
  }
  return value;
}

export function sha256Hex(payload: string): string {
  return createHash("sha256").update(payload, "utf8").digest("hex");
}

export function eventHashFromBody(body: Record<string, unknown>): string {
  return `sha256:${sha256Hex(canonicalize(body))}`;
}

File: packages/aegis-sdk-py/salanor_aegis_ledger/canonical.py
from __future__ import annotations

import hashlib
import json
from typing import Any


def _sort(value: Any) -> Any:
    if isinstance(value, list):
        return [_sort(v) for v in value]
    if isinstance(value, dict):
        return {k: _sort(value[k]) for k in sorted(value)}
    return value


def canonicalize(value: Any) -> str:
    return json.dumps(_sort(value), separators=(",", ":"), sort_keys=True)


def event_hash_from_body(body: dict[str, Any]) -> str:
    digest = hashlib.sha256(canonicalize(body).encode("utf-8")).hexdigest()
    return f"sha256:{digest}"
